sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -404,20 +404,25 @@ object ExtractSingleColumnNullAwareAntiJoin extends JoinSelectionHelper with Pre
    * which is very time consuming because it's an O(M*N) calculation.
    * But if it's a single column case O(M*N) calculation could be optimized into O(M)
    * using hash lookup instead of loop lookup.
+   *
+   * [SPARK-36665] IsNull(EqualTo(a=b)) gets optimized to Or(IsNull(a), IsNull(b))
```

Comment:
shall we limit the optimization to unary expressions only? then we don't need to change here.

Reply:
The optimization over binary expressions are powerful enough to keep it.
For example, `Or(IsNull(a), IsNull(b))` is always a better than `IsNull(EqualTo(a=b))` because `Or(IsNull(a), IsNull(b))` can go into the push down filter. In contrast, `IsNull(EqualTo(a=b))` cannot be pushed down.
It is just this specific hand-optimization pattern is outlier.
Given the benefit of optimization, small tweak like this is okay.

Reply:
I changed my mind. I took your advice and made it unaryExpressions only for now. For extending it to other expressions I will leave it as TODO.